// smile-identity-core/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMAGE_TYPE = exports.JOB_TYPE = exports.WebApi = exports.Utilities = exports.Signature = exports.IDApi = void 0;
const id_api_1 = require("./src/id-api");
Object.defineProperty(exports, "IDApi", { enumerable: true, get: function () { return id_api_1.IDApi; } });
const signature_1 = __importDefault(require("./src/signature"));
exports.Signature = signature_1.default;
const utilities_1 = require("./src/utilities");
Object.defineProperty(exports, "Utilities", { enumerable: true, get: function () { return utilities_1.Utilities; } });
const web_api_1 = require("./src/web-api");
Object.defineProperty(exports, "WebApi", { enumerable: true, get: function () { return web_api_1.WebApi; } });
const constants_1 = require("./src/constants");
Object.defineProperty(exports, "JOB_TYPE", { enumerable: true, get: function () { return constants_1.JOB_TYPE; } });
Object.defineProperty(exports, "IMAGE_TYPE", { enumerable: true, get: function () { return constants_1.IMAGE_TYPE; } });
if ((typeof process === 'undefined' &&
    typeof process.versions.node === 'undefined') ||
    typeof window !== 'undefined') {
    console.warn('This is a server-side library meant for a node.js (or compatible) runtime, and is not meant to work in the browser.');
}
